### PR TITLE
put all SNI code behind simpler preprocessor directive HAVE_SNI

### PR DESCRIPTION
**Pull request by gstrauss**
put all SNI code behind simpler preprocessor directive HAVE_SNI

(first part of #3538)

### CODE REVIEW

**Comment by kaleb-himes** [comment]
Jenkins, @gstrauss is an approved contributor. It is ok to test this PR.

- K

**Comment by gstrauss** [comment]
The definition of `wolfSSL_set_tlsext_host_name()` is hidden inside a much larger block covered by complex preprocessor logic.  To simplify this patch, I am leaving the `#if defined(WOLFSSL_NGINX) || defined(WOLFSSL_QT) || defined(OPENSSL_ALL)` around the use of `wolfSSL_set_tlsext_host_name()` in `wolfSSL_ctrl()`.

File: src/ssl.c [hunk 1]
```diff
@@ -41272,9 +41272,7 @@ int wolfSSL_get_state(const WOLFSSL* ssl)
 }
 #endif /* HAVE_LIGHTY || HAVE_STUNNEL || WOLFSSL_MYSQL_COMPATIBLE */
 
-#if defined(OPENSSL_ALL) || defined(WOLFSSL_ASIO) || defined(WOLFSSL_HAPROXY) \
-    || defined(WOLFSSL_NGINX) || defined(WOLFSSL_QT)
-
+#if defined(OPENSSL_ALL) || defined(OPENSSL_EXTRA) || defined(OPENSSL_EXTRA_X509_SMALL)
```
Comment (kaleb-himes) [inline src/ssl.c:41275]: We have customers that strictly rely on the footprint on OPENSSL_EXTRA_X509_SMALL not increasing between releases. What is the purpose for introducing new code sections to the X509_SMALL build?

**Review by kaleb-himes — CHANGES_REQUESTED** [review]
As a broad overview with the separation of `HAVE_LIGHTY, WOLFSSL_HAPROXY, WOLFSSL_NGINX, and HAVE_STUNNEL` from `HAVE_SNI` we would expect to see updates to configure.ac surrounding the `--enable-lighty, --enable-haproxy, --enable-nginx, --enable-stunnel` features to now turn on the switch `ENABLED_SNI` and set the flag `HAVE_SNI`, @dgarske can you also review when you have a moment?

File: src/ssl.c (hunk 1 above)
Comment (gstrauss) [reply src/ssl.c:41275]: The reason is described in the commit message for this commit:
"match preproc defs around wolfSSL_ctrl() in wolfssl/ssl.h"

Please indicate whether the preprocessor wrapping in the `wolfssl/ssl.h` is correct, or if the preprocessor wrapping in `src/ssl.c` is correct.

Comment (kaleb-himes) [reply src/ssl.c:41275]: Glenn,

Let's align on the ssl.c (prototype in the header doesn't increase footprint for X509_SMALL but including the function source does)

Proposed alternative:

```
diff --git a/wolfssl/ssl.h b/wolfssl/ssl.h
index b2e8a9836..5285d4456 100644
--- a/wolfssl/ssl.h
+++ b/wolfssl/ssl.h
@@ -3310,7 +3310,6 @@ WOLFSSL_LOCAL int NIDToEccEnum(int n);
 /* end of object functions */
 
 WOLFSSL_API unsigned long wolfSSL_ERR_peek_last_error_line(const char **file, int *line);
-WOLFSSL_API long wolfSSL_ctrl(WOLFSSL* ssl, int cmd, long opt, void* pt);
 WOLFSSL_API long wolfSSL_CTX_ctrl(WOLFSSL_CTX* ctx, int cmd, long opt,void* pt);
 WOLFSSL_API long wolfSSL_CTX_callback_ctrl(WOLFSSL_CTX* ctx, int cmd, void (*fp)(void));
 WOLFSSL_API long wolfSSL_CTX_clear_extra_chain_certs(WOLFSSL_CTX* ctx);
@@ -3340,6 +3339,11 @@ WOLFSSL_API int wolfSSL_check_private_key(const WOLFSSL* ssl);
 #endif /* !NO_CERTS */
 #endif /* OPENSSL_ALL || OPENSSL_EXTRA || OPENSSL_EXTRA_X509_SMALL */
 
+#if defined(OPENSSL_ALL) || defined(WOLFSSL_ASIO) || defined(WOLFSSL_HAPROXY) \
+    || defined(WOLFSSL_NGINX) || defined(WOLFSSL_QT)
+WOLFSSL_API long wolfSSL_ctrl(WOLFSSL* ssl, int cmd, long opt, void* pt);
+#endif
+
 #ifdef WOLFSSL_WPAS_SMALL
     /* WPA Supplicant requires GEN_ values */
     #include <wolfssl/openssl/x509v3.h>
```

**Comment by gstrauss** [comment]
@kaleb-himes  listed in the order that they occur in configure.ac:

`ENABLED_QT` turns on `ENABLED_SNI` and sets `-DHAVE_TLS_EXTENSIONS -DHAVE_SNI`
`ENABLED_NGINX` or `ENABLED_HAPROXY` or `ENABLED_LIGHTY` turns on `ENABLED_TLSX`
`ENABLED_TLSX` turns on `ENABLED_SNI` and sets `-DHAVE_TLS_EXTENSIONS -DHAVE_SNI`
`ENABLED_STUNNEL` turns on `ENABLED_TLSX` and sets `-DHAVE_TLS_EXTENSIONS -DHAVE_SNI`

Note: `ENABLED_TLSX` and `ENABLED_SNI` variables in configure.ac might not be enabled consistently, and might be set after the variable has been processed to update `AM_CFLAGS`.

What is important is that `-DHAVE_TLS_EXTENSIONS -DHAVE_SNI` is set in `AM_CFLAGS`.

**Comment by kaleb-himes** [comment]
@gstrauss, Thanks so much for double-checking configure.ac will set HAVE_SNI in each instance. Preliminary review is approved, over to @dgarske for final.

Warm Regards,

K

**Review by kaleb-himes — APPROVED** [review]

**Comment by kaleb-himes** [comment]
Oh I just realized the change in X509_SMALL isn't up yet, as soon as that is in I'll assign over to @dgarske for a final review.

File: src/ssl.c (hunk 1 above)
Comment (gstrauss) [reply src/ssl.c]: implemented requested change

**Comment by gstrauss** [comment]
> Known customer configs C — Failed 
looks like a spurious failure; test case does not fail for me using the same `./configure`

**Comment by gstrauss** [comment]
FYI: this PR is related to #3428

**Review by dgarske — APPROVED** [review]
These changes look good. Thanks @gstrauss

**Review by toddouska — APPROVED** [review]

**Pull request merged**